dipy/align/_public.py

```diff
@@ -0,0 +1,833 @@
+
+    Parameters
+    -----------
+    dwi : 4D array, nifti image or str
+        Containing the DWI data, or full path to a nifti file with DWI.
+    gtab : GradientTable or sequence of strings
+        The gradients associated with the DWI data, or a sequence with
+        (fbval, fbvec), full paths to bvals and bvecs files.
+    dwi_affine : 4x4 array, optional
+        An affine transformation associated with the DWI. Required if data
+        is provided as an array. If provided together with nifti/path,
+        will over-ride the affine that is in the nifti.
+    template : 3D array, nifti image or str
+        Containing the data for the template, or full path to a nifti file
+        with the template data.
+    template_affine : 4x4 array, optional
+        An affine transformation associated with the template. Required if data
+        is provided as an array. If provided together with nifti/path,
+        will over-ride the affine that is in the nifti.
+
+    reg_method : str,
+        One of "syn" or "aff", which designates which registration method is
+        used. Either syn, which uses the :func:`syn_registration` function
+        or :func:`affine_registration` function. Default: "syn".
+    reg_kwargs : key-word arguments for :func:`syn_registration` or
+        :func:`affine_registration`
+
+    Returns
+    -------
+    warped_b0, mapping: The fist is an array with the b0 volume warped to the
+    template. If reg_method is "syn", the second is a DiffeomorphicMap class
+    instance that can be used to transform between the two spaces. Otherwise,
+    if reg_method is "aff", this is a 4x4 matrix encoding the affine transform.
+
+    Notes
+    -----
+    This function assumes that the DWI data is already internally registered.
+    See :func:`register_dwi_series`.
+
+    """
+    dwi_data, dwi_affine = read_img_arr_or_path(dwi, affine=dwi_affine)
+
+    if template is None:
+        template = dpd.read_mni_template()
+
+    template_data, template_affine = read_img_arr_or_path(
+                                       template,
+                                       affine=template_affine)
+    if isinstance(dwi, str):
+        dwi = nib.load(dwi)
+
+    if not isinstance(gtab, dpg.GradientTable):
+        gtab = dpg.gradient_table(*gtab)
+
+    mean_b0 = np.mean(dwi_data[..., gtab.b0s_mask], -1)
+    if reg_method == "syn":
+        warped_b0, mapping = syn_registration(mean_b0, template_data,
+                                              moving_affine=dwi_affine,
+                                              static_affine=template_affine,
+                                              **reg_kwargs)
+    elif reg_method == "aff":
+        warped_b0, mapping = affine_registration(mean_b0, template_data,
+                                                 moving_affine=dwi_affine,
+                                                 static_affine=template_affine,
+                                                 **reg_kwargs)
```

Comment:
`else: raise an error` To avoid any strange behavior, just in case...

Reply:
Addressed below.